cpp_code/src/RandomNumberGenerator.cpp

```diff
@@ -17,28 +17,150 @@
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <stdint.h>
+
 #include "RandomNumberGenerator.h"
 
+static inline unsigned bitcount64(uint64_t x) {
+    // Count two-bit groups.
+    x -= ((x >> 1) & UINT64_C(0x5555555555555555));
+
+    // Add to four-bit groups, carefully masking carries and garbage.
+    x = ((x >> 2) & UINT64_C(0x3333333333333333)) +
+        (x & UINT64_C(0x3333333333333333));
+
+    // Add to eight-bit groups.
+    x = (((x >> 4) + x) & UINT64_C(0x0f0f0f0f0f0f0f0f));
+
+    // Add all eight-bit groups in the leftmost column of a multiply.
+    // Bit counts are small enough no other columns will carry.
+    return (x * UINT64_C(0x0101010101010101)) >> 56;
+}
+
+static inline unsigned clz64(uint64_t x) {
+    // Round up to a power of two minus one.
+    x |= x >> 1;
+    x |= x >> 2;
+    x |= x >> 4;
+    x |= x >> 8;
+    x |= x >> 16;
+    x |= x >> 32;
+
+    // Count the bits thus set, and subtract from 64 to count leading
+    // zeros.
+    return 64 - bitcount64(x);
+}
+
 //////////////////////////////////
 // return a random real between
 // 0 and 1 with uniform dist
 double RandomNumberGenerator::next() {
-    return _dist();
+    int e = -64;
+    uint64_t s;
+    unsigned d;
+
+    // Draw a significand from an infinite stream of bits.
+    while ((s = crypto_weakprng_64(&_weakprng)) == 0) {
+        e -= 64;
+        // Just return zero if the exponent is so small there are no
+        // floating-point numbers that tiny.
+        if (e < -1074)
+            return 0;
+    }
+
+    // Shift leading zeros into the exponent and fill trailing bits of
+    // significand uniformly at random.
+    if ((d = clz64(s)) != 0) {
+        e -= d;
+        s <<= d;
+        s |= crypto_weakprng_64(&_weakprng) >> (64 - d);
+    }
+
+    // Set sticky bit, since there is (almost surely) another 1 in the
+    // bit stream, to avoid a bias toward `even'.
+    s |= 1;
+
+    // Return s * 2^e.
+    return ldexp(static_cast<double>(s), e);
 }
 
 //////////////////////////////
 // return a random int bewteen
 // zero and max - 1 with uniform
 // dist if called with same max
-int RandomNumberGenerator::nexti(int max) {
-    double D = (double)max;
-    return static_cast<int>(next() * D);
+int RandomNumberGenerator::nexti(int bound) {
+    assert(0 < bound);
+    return crypto_weakprng_below(&_weakprng, bound);
+}
+
+/////////////////////////////
+// standard normal samples
+// via Box-Muller transform
+double RandomNumberGenerator::stdnormal() {
+    double u, v;
+
+    u = next();
+    v = next();
+
+    return sqrt(-2*log(u)) * sin(2*M_PI*v);
+}
+
+/////////////////////////////
+// standard Gamma samples
+//
+//       George Marsaglia & Wai Wan Tsang, `A simple method for
+//       generating gamma variables', ACM Transactions on Mathematical
+//       Software 26(3), September 2000.  DOI: 10.1145/358407.358414
+//       URI: https://dl.acm.org/citation.cfm?doid=358407.358414
```

Comment:
I wrote the code from the formulae in the paper.

Sometimes I forget that it is paywalled for non-MIT IP addresses, because I usually use an MIT IP address.

The simplified Marsaglia--Tang procedure is quoted at <https://en.wikipedia.org/wiki/Gamma_distribution#Generating_gamma-distributed_random_variables>.  I tend to prefer citations to papers in published journals, because they're typically stable for decades or centuries, and the Wikipedia article gives hardly any more information than is in the code, but I can cite Wikipedia if you would like.